src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,406 @@
+                    new_name = table.next_available_name(
+                        old_sym.name, other_table=routine_table)
+                    routine_table.rename_symbol(old_sym, new_name)
+                    table.add(old_sym)
+
+            # Check whether this symbol is used to specify the precision of
+            # any literals.
+            if old_name in precision_map:
+                for lit in precision_map[old_name]:
+                    # A literal is immutable so create a new one with the
+                    # updated symbol as its precision.
+                    dtype = ScalarType(lit.datatype.intrinsic, old_sym)
+                    lit.replace_with(Literal(lit.value, dtype))
+
+    def validate(self, node, options=None):
+        '''
+        Checks that the supplied node is a valid target for inlining.
+
+        :param node: target PSyIR node.
+        :type node: subclass of :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dict of str:values or None
+
+        :raises TransformationError: if the supplied node is not a Call.
+        :raises TransformationError: if the routine body contains a Return \
+            that is not the first or last statement.
+        :raises TransformationError: if the routine body contains a CodeBlock.
+        :raises TransformationError: if a symbol of a given name is imported \
+            from different containers at the call site and within the routine.
+        :raises TransformationError: if the routine accesses an un-resolved \
+            symbol.
+        :raises TransformationError: if a symbol declared in the parent \
+            container is accessed in the target routine.
+
+        '''
+        super().validate(node, options=options)
+
+        # The node should be a Call.
+        if not isinstance(node, Call):
+            raise TransformationError(
+                f"The target of the InlineTrans transformation "
+                f"should be a Call but found '{type(node).__name__}'.")
+
+        name = node.routine.name
+
+        # Check that we can find the source of the routine being inlined.
+        routine = self._find_routine(node)
+
+        if not routine.children or isinstance(routine.children[0], Return):
+            # An empty routine is fine.
+            return
+
+        return_stmts = routine.walk(Return)
+        if return_stmts:
+            if len(return_stmts) > 1 or not isinstance(routine.children[-1],
+                                                       Return):
+                # Either there is more than one Return statement or there is
+                # just one but it isn't the last statement of the Routine.
+                raise TransformationError(
+                    f"Routine '{name}' contains one or more "
+                    f"Return statements and therefore cannot be inlined.")
+
+        if routine.walk(CodeBlock):
```

Comment:
Similar to these, have you thought if any of the Fortran symbol attributes that end up in UnknownFortranTypes could cause problems or are we OK accepting them all?

Reply:
I guess I've partially handled this with my previous comment - named arguments are now checked for and excluded. I've added a test with a subroutine that has a pointer as argument (which works fine). Now that I write that though I realise I need to check for variables with initial values or the `SAVE` attribute.

Reply:
Actually, I've added a test for `SAVE` and it works fine. I can't think of any other potential problems at the moment.